api/samples/neoncounter.c

```diff
@@ -0,0 +1,703 @@
+static long long unsigned int count_simd_load = 0;
+static long long unsigned int count_simd_store = 0;
+static long long unsigned int count_branching = 0;
+static long long unsigned int count_taken_branches = 0;
+
+static long long unsigned int count_load = 0;
+static long long unsigned int count_load_linear = 0;
+static long long unsigned int count_load_structured = 0;
+
+static long long unsigned int count_store = 0;
+static long long unsigned int count_store_linear = 0;
+static long long unsigned int count_store_structured = 0;
+
+// use this function for clean calls
+static void
+inscount(uint num_instrs, uint num_arith, uint num_simd, uint num_simd_arith,
+         uint num_simd_load, uint num_simd_store, uint num_branching, uint num_load,
+         uint num_load_linear, uint num_load_structured, uint num_store,
+         uint num_store_linear, uint num_store_structured)
+{
+
+    count_all += num_instrs;
+    count_arith += num_arith;
+    count_simd += num_simd;
+    count_simd_arith += num_simd_arith;
+    count_simd_load += num_simd_load;
+    count_simd_store += num_simd_store;
+    count_branching += num_branching;
+    count_load += num_load;
+    count_load_linear += num_load_linear;
+    count_load_structured += num_load_structured;
+    count_store += num_store;
+    count_store_linear += num_store_linear;
+    count_store_structured += num_store_structured;
+}
+
+struct counts_t {
+    uint all;
+    uint arith;
+    uint neon;
+    uint neon_arith;
+    uint neon_load;
+    uint neon_store;
+    uint branching;
+    uint load;
+    uint load_linear;
+    uint load_structured;
+    uint store;
+    uint store_linear;
+    uint store_structured;
+};
+
+/*
+ * analyzes an instruction list
+ */
+void
+analyze_instr(instr_t *instr, void **user_data)
+{
+    // save counts for all instr. types
+    struct counts_t *counts = malloc(sizeof(struct counts_t));
+
+    counts->all = 0;
+    counts->arith = 0;
+    counts->neon = 0;
+    counts->neon_arith = 0;
+    counts->neon_load = 0;
+    counts->neon_store = 0;
+    counts->branching = 0;
+    counts->load = 0;
+    counts->load_linear = 0;
+    counts->load_structured = 0;
+    counts->store = 0;
+    counts->store_linear = 0;
+    counts->store_structured = 0;
+
+    const char *op_name;
+    int op_code;
+    // is load/store instruction
+    bool is_load = false;
+    bool is_store = false;
+    // is arithmetic instruction
+    bool is_arith = false;
+    bool is_neon = false;
+
+    instr_t *ins;
+
+    for (ins = instr; ins != NULL; ins = instr_get_next(ins)) {
+        // reset
+        is_load = false;
+        is_store = false;
+        is_arith = false;
+        is_neon = false;
+
+        // count all instructions
+        counts->all++;
+
+        if (!instr_valid(ins) || instr_is_undefined(ins)) {
+            continue;
+        }
+
+        op_code = instr_get_opcode(ins);
+        op_name = decode_opcode_name(op_code);
+
+        if (IS_DEBUG) {
```

Comment:
Is there a reason you have a dynamic check for IS_DEBUG, rather than using an ifdef? Also, do we need to decode the opcode name?

Reply:
I actually didn't think of using ifdef. I used the opcode names in an earlier version to identify instructions and still check the first letter to identify load and store instructions.